runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorker.java

```diff
@@ -848,7 +848,11 @@ public void run() {
           new TimerTask() {
             @Override
             public void run() {
-              refreshActiveWork();
+              try {
+                refreshActiveWork();
+              } catch (RuntimeException e) {
+                LOG.warn("Failed to refresh active work: ", e);
```

Comment:
Sorry I'm not that familiar with this code, could you help me understand the change? 

I'm assuming that this `refreshActiveWork()` call ultimately calls `GrpcWindmiillServier.send()`, so we're catching the error here and logging it. This is an improvement because previously that situation caused stalled GetWork streams?

Are there other places where we should be catching the `IllegalStateException`?

Reply:
Yes `refreshActiveWork` ultimately calls `GrpcWindmiillServier.send()` which could throw. This catch is to supress the exception, refreshActiveWork will get called by the timer again. 

> Are there other places where we should be catching the IllegalStateException?

Other places where `GrpcWindmiillServier.send` is called are under try catch already. 

> This is an improvement because previously that situation caused stalled GetWork streams?

This catch is not directly related to the improvement. 

The improvement is that we are now not calling `requestObserver.onNext` if `clientClosed` is true, which prevents streams stalling.   The new behavior aligns with the grpc-java doc which says don't call onNext on a closed stream https://github.com/grpc/grpc-java/blob/master/stub/src/main/java/io/grpc/stub/StreamObserver.java#L62

Reply:
Got it, thank you!